pkg/dpll/dpll.go

```diff
@@ -94,33 +129,351 @@ func NewDpll(localMaxHoldoverOffSet, localHoldoverTimeout, maxInSpecOffset int64
+		d.updateDependingProcessState(processCfg)
+		d.stateDecision(processCfg, holdoverCloseCh)
+
+	}
+	// }
+}
+
+// checks whether or not sysfs file structure exists for dpll associated with the interface
+func (d *DpllConfig) isSysFsPresent() bool {
+	path := fmt.Sprintf("/sys/class/net/%s/device/dpll_0_state", d.iface)
+	if _, err := os.Stat(path); err == nil {
+		return true
+	}
+	return false
+}
+
+// MonitorDpllNetlink monitors DPLL through netlink
+func (d *DpllConfig) MonitorDpllNetlink(processCfg config.ProcessConfig) {
+	var holdoverCloseCh chan bool
+	redial := true
+	var replies []*nl.DoDeviceGetReply
+	var err error
+	var sem *semaphore.Weighted
+	for {
+
+		// If netlink connection failed before
+		if redial {
+			if d.conn == nil {
+				conn, err := nl.Dial(nil)
+				if err != nil {
+					d.conn = nil
+					glog.Info("can't dial to DPLL netlink: ", err)
+					goto checkExit
+
+				} else {
+					d.conn = conn
+					// d.nlCtx, d.nlCtxCancel = context.WithCancel(context.Background())
+				}
+			}
+			c := d.conn.GetGenetlinkConn()
+			mcastId, found := d.conn.GetMcastGroupId(nl.DPLL_MCGRP_MONITOR)
+			if !found {
+				glog.Warning("multicast ID ", nl.DPLL_MCGRP_MONITOR, " not found")
+				goto abort
+			}
+			replies, err = d.conn.DumpDeviceGet()
+			if err != nil {
+				goto abort
+			}
+			d.nlUpdateState(replies)
+			d.updateDependingProcessState(processCfg)
+			d.stateDecision(processCfg, holdoverCloseCh)
+			err = c.JoinGroup(mcastId)
+			if err != nil {
+				goto abort
+			}
+			sem = semaphore.NewWeighted(1)
+			err = sem.Acquire(context.Background(), 1)
+			if err != nil {
+				goto abort
+			}
+
+			go func() {
+				defer sem.Release(1)
+				d.monitorNtf(c, processCfg, holdoverCloseCh)
+			}()
+			goto checkExit
+		abort:
+			d.stopDpll()
+		}
+
+	checkExit:
+		select {
+		case <-d.exitCh:
+			glog.Infof("terminating sysfs DPLL monitoring")
+			processCfg.EventChannel <- event.EventChannel{
+				ProcessName: event.DPLL,
+				IFace:       d.iface,
+				CfgName:     processCfg.ConfigName,
+				ClockType:   processCfg.ClockType,
+				Time:        time.Now().Unix(),
+				Reset:       true,
+			}
+			if d.onHoldover {
+				close(holdoverCloseCh) // cancel any holdover
+			}
+			return
+		default:
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second*2)
+			defer cancel()
+			err = sem.Acquire(ctx, 1)
+			if err != nil {
+				glog.Info("dpll monitoring is still running")
+				redial = false
+				d.updateDependingProcessState(processCfg)
+				d.stateDecision(processCfg, holdoverCloseCh)
+			} else {
+				glog.Info("dpll monitoring exited, redial")
+				d.stopDpll()
+				redial = true
+			}
+		}
+	}
+}
+
+// stopDpll stops DPLL monitoring
+func (d *DpllConfig) stopDpll() {
+	// d.nlCtxCancel()
+	d.conn.Close()
+	d.conn = nil
+}
+
+// MonitorProcess is initiating monitoring of DPLL associated with a process
 func (d *DpllConfig) MonitorProcess(processCfg config.ProcessConfig) {
+	d.processConfig = processCfg
```

Comment:
@aneeshkp I thought this line was missing somewhere, because d.processConfig remains uninitialized and used in 
```go
func (d *DpllConfig) isOffsetInRange() bool {
	if d.offset < d.processConfig.GMThreshold.Max && d.offset > d.processConfig.GMThreshold.Min {
		return true
	}
	return false
}
```
This will also allow us to remove passing processConfig as a parameter everywhere.
WDYT?